### PR TITLE
In AdoNetGrainStorage, remove AdoGrainKey and use GrainId directly

### PR DESCRIPTION
**Pull request by benjaminpetit**
Tentative to simplify `AdoNetGrainStorage` a bit.

**To avoid conflicts, #8081 needs to be merged first**

Now that all `GrainId` are essentially a pair of string, the type `AdoGrainKey` shouldn't be needed anymore. It _might_ be a bit less effective in the db to store all keys as string, but I think the simplicity it brings is worth it.

I kept the Grain key/type hashing logic. Maybe it would be worth seeing if it's still useful?

cc @veikkoeeva @ElanHasson 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/8096)

### CODE REVIEW

**Comment by ReubenBond** [comment]
I rebased this before realizing it's a Draft - apologies

**Comment by veikkoeeva** [comment]
@benjaminpetit I am not sure what you propose.

Naturally if the keys are strings only, then the integer and guid version are not needed anymore (you need to fix the [the test data sets too](https://github.com/dotnet/orleans/tree/main/test/TesterInternal/StorageTests/TestDataSets), didn't see them in change set).

The case with hashing the string identifiers is different:
- There is a database (and version?) specific limit on how many bytes the indexes can be. I could imagine the `{silo/cluster id, grain type, grain id}` combination can be too wide to be indexed. Without indexes grain storage will certainly be very slow.
- String keys in memory are a lot wider. If the indexes are narrow, and "very random", it's both easier to keep the whole index in memory and then also for the DB engine to hit they correct key and consequently resolve conflicts if any and retrieve the state.

I would suggest at least testing a bit with performance if this is the plan. Maybe script like https://github.com/dotnet/orleans/issues/5931#issuecomment-528817926 will help with that.

**Comment by benjaminpetit** [comment]
I think the test data are still fine, since you can still have `IGrainWithGuidKey`, `IGrainWithIntegerKey`, etc.

This PR keeps the hash logic, and if you think it's necessary and a dangerous change, I will not modify it. But since now all keys have a string representation, I think it's cleaner to store all keys as string and drop the int and Guid column from the database.

File: src/AdoNet/Orleans.Persistence.AdoNet/MySQL-Persistence.sql [hunk 1]
```diff
@@ -51,15 +51,13 @@ CREATE TABLE OrleansStorage
     -- The mapping is done in the code. The
     -- *String columns contain the corresponding clear name fields.
     --
-    -- If there are duplicates, they are resolved by using GrainIdN0,
-    -- GrainIdN1, GrainIdExtensionString and GrainTypeString fields.
+    -- If there are duplicates, they are resolved by using GrainIKey,
```
Comment (jonathan) [inline src/AdoNet/Orleans.Persistence.AdoNet/MySQL-Persistence.sql:54]: I believe this should read "GrainIdKey"
```suggestion
    -- If there are duplicates, they are resolved by using GrainIdKey,
```

**Review by ElanHasson — CHANGES_REQUESTED** [review]
Not sure why everything is passing.

The migrations should be run as part of the test suite from the work I did here https://github.com/dotnet/orleans/pull/7928/files#diff-363cc00495510d9c580f7d073254189014ec27981af1cd1eee28ed4c7a52b37aR54-R63

The migrations seem to be missed touchpoints here, for example this one still has `GrainIdN0` and friends:
https://github.com/dotnet/orleans/blob/f47f1737e28cada329bcc5c350ca85644e544454/src/AdoNet/Orleans.Persistence.AdoNet/Migrations/PostgreSQL-Persistence-3.6.0.sql#L13-L15


Edit:

I think it also makes sense to flatten the scripts here to a `{Provider}-{Area}.sql` file for each. I think the 3.x names will be confusing for the migrations and they should be rolled into the primary files.

**Comment by ReubenBond** [comment]
I assume this would be a breaking change now. If that's the case, do we want to take it?

**Pull request closed without merging**